Refactorings

konstin opened the pull request:
For Meine Stadt Transparent we're using exactly the same pipeline as in the example (we just copied it, to be honest), so I thought it'd be good to merge that function into the core package.

As I've already been at it, I also made refactored other parts of the codebase in the second commit, mostly applying pep8 style, but also extracting test data to files and getting rid of some try-catch-block.

torfsen commented:
Thanks for your PR, @konstin! It's nice to hear that you've found geoextract to be useful.

However, I am not sure whether moving the complete example into the core is a good idea. This would only be a benefit if it is applicable in many use cases exactly as it is configured. Since the example is written as a basic illustration I doubt that it performs well in real-life scenarios. For example, here's the normalization setup that we currently use in production:

```python
def _abbr(s):
    return r'\b{}(?:\.|\b)'.format(re.escape(s))

_SUBSTITUTIONS = [
    (_abbr('ev'), 'evangelisch'),
    (_abbr('evang'), 'evangelisch'),
    (_abbr('kath'), 'katholisch'),
    (_abbr('st'), 'sankt'),
    (_abbr('u'), 'und'),
    (_abbr('sudl'), 'sudlich'),
    (_abbr('nordl'), 'nordliche'),
    (_abbr('ostl'), 'ostliche'),
    (_abbr('westl'), 'westliche'),
    (_abbr('nsg'), 'naturschutzgebiet'),
    (_abbr('lsg'), 'landschaftsschutzgebiet'),
    (_abbr('kjt'), 'Kinder- und Jugendtreff'),
    (_abbr('kjh'), 'Kinder- und Jugendhaus'),
    (_abbr('jva'), 'Justizvollzugsanstalt'),
    (_abbr('stadt'), 'stadtisch'),
    (_abbr('staatl'), 'staatlich'),
    (_abbr('asv'), 'allgemeiner sportverein'),
    (_abbr('tc'), 'tennisclub'),
    (_abbr('e.v.'), ''),

    (r'(\w)str(?:\.|\b)', r'\1strasse'),

    ('kindertagesstatte', 'kita'),
    ('kindertageseinrichtung', 'kita'),
    ('kinderkrippe', 'kita'),
    ('betriebskindertagesstatte', 'kita'),
    ('musikkindertagesstatte', 'kita'),
    ('sportkindertagesstatte', 'kita'),
    ('kindergarten', 'kita'),
    ('schulkindergarten', 'kita'),

    ('werkrealschule', 'realschule'),

    ('ganztagesschule', 'schule'),
    ('ganztagsschule', 'schule'),
    ('ganztagsrealschule', 'realschule'),
    ('ganztagesrealschule', 'realschule'),
    ('ganztagsgrundschule', 'grundschule'),
    ('ganztagesgrundschule', 'grundschule'),
    ('ganztagsgymnasium', 'gymnasium'),
    ('ganztagesgymnasium', 'gymnasium'),

    ('kinder und jugendhaus', 'jugendhaus'),
    ('kinder und jugendtreff', 'jugendhaus'),
    ('kindertreff', 'jugendhaus'),
    ('jugendtreff', 'jugendhaus'),
    ('kinderhaus', 'jugendhaus'),

    ('jugendmusikschule', 'musikschule'),

    ('schulerhort', 'hort'),

    ('stadtteilfriedhof', 'friedhof'),

    ('ortsverwaltung', 'rathaus'),
    ('stadtamt', 'rathaus'),
    ('burgerburo', 'rathaus'),
]

_normalizer = geoextract.BasicNormalizer(subs=_SUBSTITUTIONS, stem='german')
```

As you can see, that is far more complex than what the example provides (and it's still very limited due to our focus on communal institutions). Other parts of the pipeline also require a careful configuration that is tuned to the particular use case.

I'm open to arguments for moving specific parts of the example into the core (for example, a `GermanAddressExtractor` that's basically the `PatternExtractor` as configured in the example might be a good idea).

Regarding your code style  changes: these are in general very welcome, and it's nice to see someone care about code style! However, note that PEP8 [makes no recommendation regarding single- vs. double-quotes](https://www.python.org/dev/peps/pep-0008/#id25) for normal string literals and that geoextract uses single-quotes by default. For triple-quoted literals, PEP8 recommends to follow [PEP257](https://www.python.org/dev/peps/pep-0257/), but we do not follow that recommendation (we also use a line break after the initial triple-quote).

The code style hasn't been documented before, so I took your PR as a motivation to add a linting step to our test setup. See the updated [README](https://github.com/stadt-karlsruhe/geoextract#development) for instructions.

konstin commented:
That substitution example is very educational! We're actually using the default pipeline by just throwing open street map and extracted pdfs in. We haven't done any evaluation as the locations were making a nice map. (Though we definitely should do)

Thanks for adding flake8, that makes a consistent code style much easier. I've changed the code accordingly and rebased.

I think an entrypoint that just needs a set of locations is quite useful for getting started, both when you're new to geodata extraction and for when you're getting started with a dataset. I'll do some further experimentation and make another Pull Requests when I've figured it out. Of this pull request remain mainly refactorings, so I've changed the title accordingly.

torfsen commented on example/geoextract_example.py:
```diff
@@ -2,192 +2,149 @@
 # encoding: utf-8
 
```
I'd like to keep the example *really* simple: one file, no external dependencies, just the basics to get you started. From that perspective I still don't see how your refactoring improves the example.

torfsen commented on geoextract/__init__.py:
```diff
@@ -45,7 +45,6 @@
 
 from .app import create_app
 
-
```
Interesting. I always thought that PEP8 recommends two blank lines after the imports, but I was mistaken. However, I still prefer to have two blank lines there for consistency, since often the first thing after the import section is a function or class definition (which does require two blank lines).

torfsen commented on geoextract/__init__.py:
```diff
@@ -205,7 +205,9 @@ def extract(self, text):  # noqa: D102
-            yield (end_index - length + 1, length, {'name': name})
+            yield (end_index - length + 1, length, {
+                'name': name
+            })
```
That definitely doesn't improve readability for me :wink:

torfsen commented on geoextract/__init__.py:
```diff
@@ -434,10 +437,9 @@ def _augment_result(self, result):
-                pass
+            if key not in result or result[key] not in self.normalized_names:
+                continue
+            result[key] = self.normalized_names[result[key]]['name']
```
In Python, exceptions are used differently than in many other languages (see, for example, [this post](https://softwareengineering.stackexchange.com/a/351121/233918)). In particular, Python follows the ["ask for forgiveness, not permission" paradigm](https://docs.quantifiedcode.com/python-anti-patterns/readability/asking_for_permission_instead_of_forgiveness_when_working_with_files.html). This means that catching exceptions is preferred to a priori checks. Hence, this change and the other similar ones are not an improvement.

torfsen commented on geoextract/app.py:
```diff
@@ -57,9 +57,9 @@ def extract():
             text = request.files['text'].read().decode('utf-8')
         except KeyError:
-            abort(400, 'Missing "text" parameter.')
+            return abort(400, 'Missing "text" parameter.')
```
According to the [documentation](http://flask.pocoo.org/docs/0.12/api/#flask.abort), `abort` raises an exception, so there's no point in adding a `return`.

torfsen commented on geoextract/__init__.py:
```diff
@@ -594,7 +595,7 @@ class BasicNormalizer(Normalizer):
 
     def __init__(self, to_ascii=True, rejoin_lines=True, remove_hyphens=True,
-                 remove_specials=True, subs=None, stem=None):
+                 remove_specials=True, substitutions=None, stem=None):
```
I agree that `substitutions` is the better name. However, given the method's documentation the benefit of the change is small and does IMO not justify breaking backwards-compatibility by changing the public interface of the class.

konstin commented:
I guess I've been too quick about just changing things because they felt right. As there's not real value remaining, I'm going to close this pull request.

For my own code I've written an AddressPipeline class based on the example code. I'll probably do some manual evaluation and build a substitution list.

```python
class AddressPipeline(geoextract.Pipeline):
    def __init__(self, locations, subs=None, stem='german'):
        if subs is None:
            if stem == 'german':
                subs = [(r'str\b', 'strasse')]
            else:
                subs = []
        normalizer = geoextract.BasicNormalizer(subs=subs, stem=stem)

        name_extractor = geoextract.NameExtractor()

        address_pattern = re.compile(r'''
            (?P<street>[^\W\d_](?:[^\W\d_]|\s)*[^\W\d_])
            \s+
            (?P<house_number>([1-9]\d*)[\w-]*)
            (
                \s+
                (
                    (?P<postcode>\d{5})
                    \s+
                )?
                (?P<city>([^\W\d_]|-)+)
            )?
        ''', flags=re.UNICODE | re.VERBOSE)

        pattern_extractor = geoextract.PatternExtractor([address_pattern])

        extractors = [pattern_extractor, name_extractor]

        keys_to_keep = ['name', 'street', 'house_number', 'postcode', 'city']
        postprocessors = [(geoextract.KeyFilterPostprocessor(keys_to_keep))]

        super().__init__(locations,
                         extractors=extractors,
                         normalizer=normalizer,
                         postprocessors=postprocessors)
```

torfsen commented:
@konstin Thanks for sharing your `AddressPipeline`! Feel free to share your substitution list, too, once you're satisfied with it, I'm certain you'll find things that are of interest to us (and others), too.
